### ISSUE TITLE
Signing messages with a custom "types" in the CCXT library

### ISSUE THREAD

**Issue by allexx9** [ccxt/ccxt#6385]
Hi guys, i'm working on the implementation one the connectors to the DEX exchange, and facing with a interesting problem. For submit orders to our exchange we need to sign a request types/values by Keccak/Sha3. Sha3 implementation in the Exchange class in the CCXT is generate a types for values automatically, but for correct hash (in the side our exchange) we need to use a uint8 instead of uint256 in the one of request parameters. 
In the Exchange class for get hashes with custom types you use a ethAbi.solidityKeccak(), but in the my market implementation i can't get access to this method from Exchange class. 
How can I access this function, or what alternatives can I use to implement this using the internal CCXT functionality, without using external dependencies ?

**Comment by kroitor** [comment]
Hi! Can you paste a snippet of code for some context? Which particular element in the order structure should be `uint8` in your case?

**Comment by allexx9** [comment]
Hi, thanks for answering, basically i want to do something like:
```
generateOrderHash (request) {
        return this.solidityKeccak ([
            'address',
            'uint8',
            'address',
            'address',
            'address',
            'address',
            'uint256',
            'uint256',
            'uint256',
            'uint256',
            'uint256',
            'uint256',
            'uint256',
        ], [
            this.web3.utils.toChecksumAddress (request['exchangeContractAddress'].toLowerCase ()),
            request['side'],
            this.web3.utils.toChecksumAddress (request['address'].toLowerCase ()),
            this.web3.utils.toChecksumAddress (request['tokenBaseAddress'].toLowerCase ()),
            this.web3.utils.toChecksumAddress (request['tokenQuoteAddress'].toLowerCase ()),
            this.web3.utils.toChecksumAddress (request['tokenFeeAddress'].toLowerCase ()),
            request['amount'],
            request['priceNumerator'],
            request['priceDenominator'],
            request['feeNumerator'],
            request['feeDenominator'],
            request['expirationTs'],
            request['nonce'],
        ]);
    }
```
if i use exchange.    soliditySha3 () for it - type of 'side' value defined as uint256.

**Comment by kroitor** [comment]
@allexx9 yep, it may be tricky... Also, we can't use toChecksumAddress without porting `web3.utils` to PHP. We will look into adding a generalization for the above use case to make those methods reusable and configurable where necessary.

**Comment by allexx9** [comment]
Thanks! Then i can't correctly get a hash? btw, maybe you have any list of Web3 functions which i can safely use for the process which i described? (Sha3/Keccak from Web3 directly, and something to replace my calls to toChecksumAddress?)

**Comment by kroitor** [comment]
Yes, we will add some edits for that and will document those methods shortly. Hope to get to it asap.

**Comment by allexx9** [comment]
Thanks!

**Comment by frosty00** [comment]
yeah currently we don't have support for all the solidity types, just uint256, address (and string in php)

**Comment by carlotestor** [comment]
Updating this long-standing thread: CCXT now has proper ABI encoding built in, so custom Solidity types including `uint8` are supported. Use `ethAbiEncode` (full ABI encoding with explicit types) together with `keccak`:

```javascript
generateOrderHash (request) {
    const encoded = this.ethAbiEncode ([
        'address', 'uint8', 'address', 'address', 'address', 'address',
        'uint256', 'uint256', 'uint256', 'uint256', 'uint256', 'uint256', 'uint256',
    ], [
        request['exchangeContractAddress'],
        request['side'],                    // uint8 - now supported
        request['address'],
        request['tokenBaseAddress'],
        request['tokenQuoteAddress'],
        request['tokenFeeAddress'],
        request['amount'],
        request['priceNumerator'],
        request['priceDenominator'],
        request['feeNumerator'],
        request['feeDenominator'],
        request['expirationTs'],
        request['salt'],
    ]);
    return this.hash (encoded, keccak, 'hex');
}
```

Also relevant to your follow-up questions:
- `this.ethEncodeStructuredData (domain, types, data)` covers EIP-712 typed-data signing.
- `this.ethGetAddressFromPrivateKey (pk)` derives the address, so `toChecksumAddress` from web3.utils is no longer needed for that purpose.
- These are all transpiled to Python/PHP, so no external web3 dependency is required in an exchange implementation.

`hyperliquid`, `paradex`, `derive` and `dydx` in the repo are good working references for the signing patterns. If you're still building that connector, a PR would be welcome.
